package f3b;
import public "f3b/types.proto";

message PhysicsData {

    enum PhysicsShape {
        smesh = 1;
        ssphere = 2;
        shull = 3;
        sbox = 4;
        scapsule = 5;
        scylinder = 6;
        scone = 7;
    }
   
    optional RigidBody rigidbody = 1;
    optional Constraint constraint = 2;
}   

message Constraint {
    required string id = 1;
    required string a_ref = 2;
    required string b_ref = 3;
    optional ConstraintGeneric generic = 4;   

}

message ConstraintGeneric {
    required vec3 pivotA = 1;
    required vec3 pivotB = 2;
    required vec3 upperLinearLimit = 3;
    required vec3 lowerLinearLimit = 4;
    required vec3 upperAngularLimit = 5;
    required vec3 lowerAngularLimit = 6;
    required bool disable_collisions = 7;

}
   
message RigidBody {
    required string id = 1;

    enum RigidBodyType {
        tnone = 1;
        tstatic = 2;
        tdynamic = 3;
        tghost = 4;
    }
    
    optional RigidBodyType type = 2 [ default = tnone ];
    optional PhysicsData.PhysicsShape shape = 3 [ default = smesh ];
    optional float mass = 4 [ default = 1.0 ];
    optional float friction = 5 [ default = 1.0 ];
    optional float angularDamping = 6 [ default = 0 ];
    optional float linearDamping = 7 [ default = 0 ];
    optional float margin = 8 [ default = 0 ];
    optional float restitution = 9 [ default = 0 ];
    optional vec3 angularFactor = 10;
    optional vec3 linearFactor = 11; 
    optional bool isKinematic = 12 [ default=false ];
    optional int32 collisionGroup = 13 [ default = 1 ];
    optional int32 collisionMask = 14 [ default = 1 ];
    
}
